src/Components/Server/src/Circuits/CircuitFactory.cs

```diff
@@ -1,18 +1,114 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
+using System.Collections.Generic;
+using System.Linq;
 using System.Security.Claims;
-using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Components.Routing;
+using Microsoft.AspNetCore.Components.Web.Rendering;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.JSInterop;
 
 namespace Microsoft.AspNetCore.Components.Server.Circuits
 {
-    internal abstract class CircuitFactory
+    internal class CircuitFactory
     {
-        public abstract CircuitHost CreateCircuitHost(
-            HttpContext httpContext,
+        private readonly IServiceScopeFactory _scopeFactory;
+        private readonly ILoggerFactory _loggerFactory;
+        private readonly CircuitIdFactory _circuitIdFactory;
+        private readonly CircuitOptions _options;
+        private readonly ILogger _logger;
+
+        public CircuitFactory(
+            IServiceScopeFactory scopeFactory,
+            ILoggerFactory loggerFactory,
+            CircuitIdFactory circuitIdFactory,
+            IOptions<CircuitOptions> options)
+        {
+            _scopeFactory = scopeFactory ?? throw new ArgumentNullException(nameof(scopeFactory));
+            _loggerFactory = loggerFactory ?? throw new ArgumentNullException(nameof(loggerFactory));
+            _circuitIdFactory = circuitIdFactory ?? throw new ArgumentNullException(nameof(circuitIdFactory));
+            _options = options?.Value ?? throw new ArgumentNullException(nameof(options));
```

Comment:
Suggest getting rid of this `?.` It's really easy to use options in tests.

Reply:
Got rid of all the null checking in here. We don't need it

Reply:
There are a lot of things we don't need in life 😆 Shoes, iced coffee, iguanas...

I've been in the habit of putting null checks in constructors because it keeps people from getting lazy in tests. There were a bunch of cases where I've seen people (in general, not pointing a finger at you) passing null into constructors of things in tests - because on the code path that they're testing it's not required. This usually blows up in someone's face when they have to refactor the code.